codalab/worker/singularity_image_manager.py

```diff
@@ -0,0 +1,95 @@
+import logging
+import os
+
+from codalab.worker.image_manager import ImageAvailabilityState
+from codalab.worker.fsm import DependencyStage
+from codalab.worker.image_manager import ImageManager
+from spython.main import Client
+
+logger = logging.getLogger(__name__)
+
+
+class SingularityImageManager(ImageManager):
+    def __init__(self, max_image_size: int, max_image_cache_size: int, image_folder: str):
+        """
+
+        Args:
+            max_image_size: Maximum image size of any pulled singularity image.
+            max_image_cache_size: The maximum size of the image cache (folder images are pulled into)
+            image_folder: folder images are pulled into.
+        """
+        super().__init__(max_image_size, max_image_cache_size)
+        if not os.path.isdir(image_folder):
+            raise ValueError("image_folder %s is not a directory" % image_folder)
+        self.image_folder = image_folder
+
+    def _cleanup(self):
+        files = os.listdir(self.image_folder)
+        for f in files:
+            os.remove(os.path.join(self.image_folder, f))
+
+    def get(self, image_spec):
+        """
+        image_spec is in singularity format:
+        - Docker: 'docker://<image>'
+        - Sylabs Cloud hub: 'library://<image>'
+        - Singularity Hub (deprecated): 'shub://<image>'
+        Images will be assumed to be docker if no prefix is included.
+        """
+        if len(image_spec.split("://")) <= 1:
+            # prefix docker, no scheme exists
+            image_spec = "docker://" + image_spec
```

Comment:
i think its better to keep the `:latest` appendation where it is.

We shouldn't modify the `image_spec` in `image_manager.py` because theres no way to tell whether there is a version for singularity (the scheme (`docker://`) contains a colon)

Reply:
> We shouldn't modify the `image_spec` in `image_manager.py` because theres no way to tell whether there is a version for singularity (the scheme (`docker://`) contains a colon)

There is, right? You could strip the scheme first and then check if there's a colon. Versions will not look like `://`.

Reply:
yeah. fixed